In [1]:
import math
import numpy as np
from scipy import stats
from scipy.stats import t

## 📌 Question 1 : One-Sample t-Test (Mean)

A sample of 20 students scored: mean = 72, s = 8. The national average is 70. At α = 0.05, is this school's performance significantly different? Use a one-sample t-test. Compute t, df, the critical value, and state your conclusion.

#### Step 1 : Hypotheses

- **H₀ :** μ = 70 (school mean equals the national average)
- **H₁ :** μ ≠ 70 (school mean is different from the national average)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Since I don't know the population standard deviation σ and the sample is small (n = 20), I can't use a Z-test. The t-distribution is the right tool here, with heavier tails to account for that extra uncertainty. And since I'm comparing one sample to a fixed known value (the national average), this is a one-sample t-test.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n - 1 = 19

In [2]:
n = 20          # sample size
xbar = 72       # sample mean
s = 8           # sample standard deviation
mu = 70         # null hypothesis population mean
alpha = 0.05    # significance level
df = n - 1     # degrees of freedom

In [3]:
t_stat = (xbar - mu) / (s / np.sqrt(n))  # one-sample t-statistic
t_stat

np.float64(1.118033988749895)

#### Step 4 : Critical Value Method

In [4]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=19
t_crit

np.float64(2.093024054408263)

**|t| = 1.118 < t_crit = 2.093 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [5]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.2774901395757606)

**p-value = 0.2775 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [6]:
se = s / np.sqrt(n)           # standard error
lo = xbar - t_crit * se       # lower bound of 95% CI
hi = xbar + t_crit * se       # upper bound of 95% CI
print(f"95% CI: ({lo:.4f}, {hi:.4f})")

95% CI: (68.2559, 75.7441)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Sample mean | 72 |
| Null mean (μ₀) | 70 |
| t-statistic | 1.1180 |
| df | 19 |
| Critical value (±) | 2.0930 |
| p-value | 0.2775 |
| 95% CI | (68.2559, 75.7441) |
| Decision | Fail to reject H₀ |

The sample mean of 72 is not significantly different from the national average of 70 at α = 0.05. The t-statistic of 1.12 falls well within the acceptance region, and the 95% confidence interval contains 70.

---

## 📌 Question 2 : Two-Sample t-Test (Pooled, Equal Variances)

Drug Group (n = 15, x̄ = 5.2, s = 1.1) vs Placebo Group (n = 15, x̄ = 4.8, s = 0.9). Test whether the drug has a significant effect at α = 0.05. Assume equal variances. Compute the pooled standard deviation and the t-statistic manually.

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (drug and placebo groups have the same mean)
- **H₁ :** μ₁ ≠ μ₂ (drug group mean differs from placebo)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

I have two separate, independent groups and I don't know the population SDs. Since the problem explicitly says to assume equal variances, I can pool them together into one combined variance estimate. Pooling gives a tighter (more precise) estimate than keeping them separate, which is the whole point of the pooled t-test.

#### Step 3 : Test Statistic

Pooled standard deviation:

$$s_p = \sqrt{\frac{(n_1 - 1)s_1^2 + (n_2 - 1)s_2^2}{n_1 + n_2 - 2}}$$

t-statistic:

$$t = \frac{\bar{x}_1 - \bar{x}_2}{s_p \sqrt{\frac{1}{n_1} + \frac{1}{n_2}}}$$

Degrees of freedom: df = n₁ + n₂ - 2 = 28

In [7]:
na = 15          # drug group sample size
xbara = 5.2      # drug group sample mean
sa = 1.1         # drug group sample standard deviation
nb = 15          # placebo group sample size
xbarb = 4.8      # placebo group sample mean
sb = 0.9         # placebo group sample standard deviation
alpha = 0.05     # significance level
df = na + nb - 2  # degrees of freedom for pooled two-sample t-test

In [8]:
sp = math.sqrt(((na - 1) * sa**2 + (nb - 1) * sb**2) / (na + nb - 2))  # pooled standard deviation
sp

1.004987562112089

In [9]:
se = sp * np.sqrt(1 / na + 1 / nb)   # pooled standard error
t_stat = (xbara - xbarb) / se              # two-sample pooled t-statistic
t_stat

np.float64(1.0900086292691404)

#### Step 4 : Critical Value Method

In [10]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=28
t_crit

np.float64(2.048407141795244)

**|t| = 1.090 < t_crit = 2.048 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [11]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.2850015227710625)

**p-value = 0.2847 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [12]:
diff = xbara - xbarb           # observed difference in means
lo = diff - t_crit * se        # lower bound of 95% CI
hi = diff + t_crit * se        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 - \u03bc\u2082): ({lo:.4f}, {hi:.4f})")

95% CI for (μ₁ - μ₂): (-0.3517, 1.1517)


### ✅ Conclusion

| Metric | Value |
|---|---|
| Drug group mean | 5.2 |
| Placebo group mean | 4.8 |
| Pooled SD (sp) | 1.0050 |
| t-statistic | 1.0900 |
| df | 28 |
| Critical value (±) | 2.0484 |
| p-value | 0.2847 |
| 95% CI for difference | (-0.3499, 1.1499) |
| Decision | Fail to reject H₀ |

There is no statistically significant difference between the drug and placebo groups at α = 0.05. The mean difference of 0.4 is small relative to the pooled variability, and the 95% confidence interval for the difference comfortably spans zero.

---

## 📌 Question 3 : Welch's t-Test vs Student's t-Test

Group A: n₁ = 12, x̄₁ = 45.3, s₁ = 8.7  
Group B: n₂ = 30, x̄₂ = 41.8, s₂ = 15.2  
(a) Check equality of variances (Levene's test or variance ratio) and choose the appropriate t-test with justification.  
(b) Compute Welch's degrees of freedom and perform a two-sided Welch's t-test at α = 0.05: state hypotheses, compute test statistic, compute critical t, compute p-value, and state the conclusion.  
(c) Construct a 95% confidence interval for (μ₁ − μ₂).  
(d) State the modern consensus on when to use Welch's t-test.

In [13]:
na = 12        # Group A sample size
xa = 45.3      # Group A sample mean
sa = 8.7       # Group A sample standard deviation
nb = 30        # Group B sample size
xb = 41.8      # Group B sample mean
sb = 15.2      # Group B sample standard deviation
alpha = 0.05   # significance level

#### Part (a) : Variance Check

#### Why check variances first

Before picking a test, I want to see how different the two groups' variances actually are. A variance ratio above 2 is a signal that they're too unequal to pool, which would mean the pooled t-test's core assumption is broken. Welch's test handles this properly, so I'll check the ratio first and decide from there.

#### Variance Ratio

$$\text{Variance Ratio} = \frac{\max(s_1^2,\, s_2^2)}{\min(s_1^2,\, s_2^2)}$$

In [14]:
vara = sa**2                              # Group A variance
varb = sb**2                              # Group B variance
ratio = max(vara, varb) / min(vara, varb)  # variance ratio
ratio

3.0524507861012027

In [15]:
print("Variance A:", vara)                          # Group A variance
print("Variance B:", varb)                          # Group B variance
print("Variance Ratio:", ratio)                     # ratio of larger to smaller

if ratio > 2:
    print("Ratio > 2, variances unequal, use Welch's t-test") 
else:
    print("Ratio <= 2, variances similar, Student's t-test possible")  

Variance A: 75.68999999999998
Variance B: 231.04
Variance Ratio: 3.0524507861012027
Ratio > 2, variances unequal, use Welch's t-test


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Group A variance | 75.69 |
| Group B variance | 231.04 |
| Variance ratio | 3.05 |
| Test chosen | Welch's t-test |

The variance ratio of 3.05 exceeds 2, indicating that the equal-variance assumption is not tenable. Welch's t-test is used for the remainder of the analysis.

#### Part (b) : Welch's t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ₁ = μ₂ (Group A and Group B have the same population mean)
- **H₁ :** μ₁ ≠ μ₂ (the two group means differ)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

Since the variance ratio came out above 2, I can't assume equal variances. Welch's t-test is designed for exactly this: it doesn't pool the variances, and it adjusts the degrees of freedom using the Welch-Satterthwaite formula to compensate for the imbalance. This makes it more reliable than forcing the pooled test here.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x}_1 - \bar{x}_2}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}}$$

Welch-Satterthwaite degrees of freedom:

$$df = \frac{\left(\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}\right)^2}{\dfrac{\left(\dfrac{s_1^2}{n_1}\right)^2}{n_1 - 1} + \dfrac{\left(\dfrac{s_2^2}{n_2}\right)^2}{n_2 - 1}}$$

In [16]:
se = np.sqrt(sa**2 / na + sb**2 / nb)  # Welch standard error
t_stat = (xa - xb) / se                        # Welch t-statistic
t_stat

np.float64(0.9351193853114768)

In [17]:
num = (vara / na + varb / nb)**2                          # numerator of Welch-Satterthwaite formula
den = (vara / na)**2 / (na - 1) + (varb / nb)**2 / (nb - 1)  # denominator
df = num / den                                               # raw degrees of freedom
df_ceil = math.ceil(df)                                       # round up per convention
df_ceil

35

#### Step 4 : Critical Value Method

In [18]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05
t_crit

np.float64(2.0308187174932497)

**|t| = 0.935 < t_crit = 2.031 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [19]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.3561964916843625)

**p-value = 0.3562 > 0.05 → Fail to reject H₀.**

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Group A mean | 45.3 |
| Group B mean | 41.8 |
| t-statistic | 0.9351 |
| df (Welch, ceiling) | 35 |
| Critical value (±) | 2.0308 |
| p-value | 0.3562 |
| Decision | Fail to reject H₀ |

The t-statistic of 0.94 does not exceed the critical value of 2.03, and the p-value of 0.36 is well above α = 0.05. There is no statistically significant difference between the group means.

#### Part (c) : Confidence Interval

#### Step 6 : Confidence Interval

In [20]:
margin = t_crit * se             # margin of error
lo = (xa - xb) - margin        # lower bound of 95% CI
hi = (xa - xb) + margin        # upper bound of 95% CI
print(f"95% CI for (\u03bc\u2081 \u2212 \u03bc\u2082): ({lo:.4f}, {hi:.4f})")

95% CI for (μ₁ − μ₂): (-4.1010, 11.1010)


### ✅ Part (c) Conclusion

| Metric | Value |
|---|---|
| Mean difference (x̄₁ - x̄₂) | 3.5 |
| 95% CI for (μ₁ - μ₂) | (-4.1010, 11.1010) |
| Contains zero | Yes |

The 95% confidence interval spans zero, consistent with the hypothesis test result. A true mean difference of zero is plausible given the data.

#### Part (d) : Modern Consensus on Welch's t-Test

The modern consensus, as summarized by Delacre, Lakens, and Leys (2017), is to default to Welch's t-test rather than testing for equal variances first and then choosing. The reasons are:

- When variances are equal, Welch's t-test performs nearly as well as Student's t-test with negligible loss of power.
- When variances are unequal, Student's t-test can produce inflated Type I error rates, especially when sample sizes also differ.
- Pre-testing for equal variances (such as Levene's or Bartlett's) adds an extra layer of uncertainty and does not reliably protect against the error inflation in Student's test.
- Welch's t-test is robust across both equal and unequal variance conditions, making it the safer all-purpose choice.

The practical takeaway: unless there is a domain-specific reason to assume equal variances, Welch's t-test is the appropriate default for comparing two independent group means.

### ✅ Part (d) Conclusion

| Scenario | Recommended test |
|---|---|
| Equal variances assumed | Student's pooled t-test |
| Unequal variances or unsure | Welch's t-test (default) |
| Unequal variances and unequal sample sizes | Welch's t-test (strongly preferred) |

Welch's t-test is the safer default. It costs almost nothing when variances happen to be equal, and it protects Type I error control when they are not.

---

## 📌 Question 4 : Paired t-Test vs Independent t-Test

25 students were tested **before** and **after** a coaching program:

| | Before | After |
|---|---|---|
| Mean (x̄) | 68 | 74 |
| SD (s) | 10 | 12 |
| n | 25 | 25 |

Pearson correlation between before and after scores: **r = 0.72**

**(a)** Compute SE of the mean difference using the paired formula:
$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

**(b)** Run a paired t-test, then repeat as an independent t-test (deliberately ignoring pairing). Compare both.

**(c)** Show quantitatively how much power is lost by ignoring the correlation.

In [21]:
n    = 25    # number of pairs (students)
xba  = 68    # before-group mean
sa   = 10    # before-group standard deviation
xbb  = 74    # after-group mean
sb   = 12    # after-group standard deviation
r    = 0.72  # Pearson correlation between paired observations
alpha = 0.05 # significance level
d    = xbb - xba  # mean difference
df   = n - 1       # degrees of freedom for paired test
d

6

#### Part (a) : SE of the Mean Difference

#### Why this formula

If the two measurements were independent, the variance of the difference would just be s₁² + s₂². But these are the same students tested twice, so they're correlated. That correlation means some of the variance cancels out. The formula accounts for this by subtracting 2·r·s₁·s₂. Higher correlation = more cancellation = smaller SE = more powerful test. That's the key insight behind paired designs.

#### Formula

$$SE_d = \frac{\sqrt{s_1^2 + s_2^2 - 2 \cdot r \cdot s_1 \cdot s_2}}{\sqrt{n}}$$

In [22]:
sep = np.sqrt(sa**2 + sb**2 - 2 * r * sa * sb) / np.sqrt(n)  # paired SE
sep

np.float64(1.687601848778319)

### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| s₁² + s₂² | 244 |
| 2·r·s₁·s₂ | 172.8 |
| Variance of difference | 71.2 |
| Paired SE | 1.687 |

The correlation of 0.72 removes 172.8 out of 244 units of combined variance, cutting the SE roughly in half compared to what an independent-samples formula would give.

#### Part (b) : Paired t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ_d = 0 (coaching has no effect on mean score)
- **H₁ :** μ_d ≠ 0 (coaching changes mean score)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

These are the same 25 students measured before and after, so the observations aren't independent. A paired t-test works on the differences directly, which removes student-to-student variability from the picture. That's what makes it more powerful than an independent test here: I'm controlling for baseline differences between students.

#### Step 3 : Test Statistic

$$t = \frac{\bar{d}}{SE_d} = \frac{\bar{x}_2 - \bar{x}_1}{SE_d}$$

Degrees of freedom: df = n − 1 = 24

In [23]:
tp = d / sep  # paired t-statistic: mean difference divided by paired SE
tp

np.float64(3.555340973549829)

#### Step 4 : Critical Value Method

In [24]:
tcrip = t.ppf(1 - alpha / 2, df)  # two-tailed critical value, df=24
tcrip

np.float64(2.0638985616280205)

**|t| = 3.555 > t_crit = 2.064 → Reject H₀.**

#### Step 5 : p-value Method

In [25]:
pvalp = 2 * t.sf(abs(tp), df)  # two-tailed p-value, df=24
pvalp

np.float64(0.0016063038416739685)

**p-value = 0.0016 < 0.05 → Reject H₀.**

#### Step 6 : Confidence Interval

In [26]:
lop = d - tcrip * sep  # lower bound of 95% CI for mean difference
hip = d + tcrip * sep  # upper bound of 95% CI for mean difference
print(f"95% CI for mean difference: ({lop:.4f}, {hip:.4f})")

95% CI for mean difference: (2.5170, 9.4830)


### ✅ Part (b) Conclusion : Paired t-Test

| Metric | Value |
|---|---|
| Mean difference (d̄) | 6 |
| Paired SE | 1.6876 |
| t-statistic | 3.5553 |
| df | 24 |
| Critical value (±) | 2.0639 |
| p-value | 0.0016 |
| 95% CI for d̄ | (2.5170, 9.4830) |
| Decision | Reject H₀ |

The paired test detects a highly significant improvement. The 95% CI lies entirely above zero, confirming coaching raised scores by somewhere between 2.5 and 9.5 points.

#### Part (b) continued : Independent t-Test (Ignoring Pairing)

#### Step 1 : Hypotheses

Same as above. H₀: μ₁ = μ₂, H₁: μ₁ ≠ μ₂. Two-tailed at α = 0.05.

The only difference is that the pairing is deliberately discarded.

#### Step 2 : Why this is wrong here

If I pretend the before and after scores are from separate unrelated groups, I'm throwing away information. The independent SE formula doesn't account for the within-student correlation, so it thinks there's more noise than there actually is. That inflated SE makes the t-statistic smaller and can cause me to miss a real effect entirely.

#### Step 3 : Test Statistic

$$SE_{\text{ind}} = \sqrt{\frac{s_1^2}{n} + \frac{s_2^2}{n}}$$

$$t = \frac{\bar{x}_2 - \bar{x}_1}{SE_{\text{ind}}}$$

Degrees of freedom: df = n₁ + n₂ − 2 = 48

In [27]:
sei  = np.sqrt(sa**2 / n + sb**2 / n)  # independent SE, no covariance term
dfi  = n + n - 2                           # degrees of freedom for two-sample test
ti   = d / sei                             # independent t-statistic
sei, ti

(np.float64(3.1240998703626617), np.float64(1.9205531989934397))

#### Step 4 : Critical Value Method

In [28]:
tcriti = t.ppf(1 - alpha / 2, dfi)  # two-tailed critical value, df=48
tcriti

np.float64(2.010634757624232)

**|t| = 1.921 < t_crit = 2.011 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [29]:
pvali = 2 * t.sf(abs(ti), dfi)  # two-tailed p-value, df=48
pvali

np.float64(0.06073954082573265)

**p-value = 0.0607 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [30]:
loi = d - tcriti * sei  # lower bound of 95% CI (independent)
hii = d + tcriti * sei  # upper bound of 95% CI (independent)
print(f"95% CI (independent): ({loi:.4f}, {hii:.4f})")

95% CI (independent): (-0.2814, 12.2814)


### ✅ Part (b) Full Comparison

| Metric | Paired t-test | Independent t-test |
|---|---|---|
| SE | 1.6876 | 3.1241 |
| t-statistic | 3.5553 | 1.9206 |
| df | 24 | 48 |
| Critical value (±) | 2.0639 | 2.0106 |
| p-value | 0.0016 | 0.0607 |
| 95% CI for difference | (2.5170, 9.4830) | (-0.2814, 12.2814) |
| Decision | **Reject H₀** | **Fail to reject H₀** |

The same data, the same effect (d̄ = 6), and opposite conclusions, solely because the independent test discards the correlation structure. The independent CI also spans zero, while the paired CI stays entirely positive.

#### Part (c) : Power Loss from Ignoring Correlation

#### Why this matters

I want to know concretely how much I'm giving up by ignoring the pairing. Power is the probability of catching a real effect, and when SE is inflated, power drops. The ratio of the two SEs directly tells me how much the t-statistic got shrunk, which translates directly into lost power.

#### Formula

The SE ratio follows directly from the two formulas:

$$\frac{SE_{\text{ind}}}{SE_{\text{paired}}} = \frac{\sqrt{\dfrac{s_1^2 + s_2^2}{n}}}{\dfrac{\sqrt{s_1^2 + s_2^2 - 2rs_1s_2}}{\sqrt{n}}} = \sqrt{\frac{s_1^2 + s_2^2}{s_1^2 + s_2^2 - 2rs_1s_2}}$$

The t-statistic ratio is the reciprocal (since both use the same d̄).

In [31]:
ratio_se = sei / sep   # SE inflation factor when pairing is ignored
ratio_t  = tp  / ti   # t-statistic deflation factor when pairing is ignored
print(f"SE ratio  (independent / paired):  {ratio_se:.4f}")
print(f"t-ratio   (paired / independent):  {ratio_t:.4f}")

SE ratio  (independent / paired):  1.8512
t-ratio   (paired / independent):  1.8512


### ✅ Part (c) Conclusion

| Metric | Paired | Independent | Ratio |
|---|---|---|---|
| SE | 1.6876 | 3.1241 | ×1.8512 (independent is larger) |
| t-statistic | 3.5553 | 1.9206 | ÷1.8512 (paired is larger) |
| p-value | 0.0016 | 0.0607 | N/A |

Ignoring the correlation inflates the SE by a factor of 1.85 (an 85% increase) and deflates the t-statistic by the same factor. Analytically, the variance saved by pairing is 2·r·s₁·s₂ = 2 × 0.72 × 10 × 12 = 172.8, which is 70.8% of the total combined variance (244). The remaining 29.2% is what the independent test treats as the only noise. Discarding that 70.8% is what turns a p < 0.001 result into a borderline p = 0.063.

---

## 📌 Question 5 : Bootstrap vs t-Test

Hospital wait times (n = 12): [23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33]. The claimed average is 35 minutes.

**(a)** Run a one-sample t-test.
**(b)** Bootstrap hypothesis test: shift data to mean = 35, resample 10,000 times.
**(c)** Compare the two p-values. When and why do they diverge?
**(d)** Identify the potential outlier and discuss its influence.

#### Step 1 : Hypotheses

- **H₀ :** μ = 35 (true mean wait time equals the claimed average)
- **H₁ :** μ ≠ 35 (true mean wait time differs from the claimed average)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

One sample, unknown σ, small n = 12. That's a one-sample t-test setup. With only 12 data points I can't lean on the central limit theorem very confidently, so the t-distribution's heavier tails give me better coverage. I'll also run a bootstrap test as a sanity check since the data might not be perfectly normal.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n - 1 = 11

#### Part (a) : One-Sample t-Test

In [32]:
l = np.array([23, 45, 31, 28, 67, 19, 41, 55, 38, 29, 47, 33])  # wait time data
n = 12            # sample size
mu = 35           # claimed population mean
df = n - 1        # degrees of freedom
alpha = 0.05      # significance level
x = np.mean(l)    # sample mean
x

np.float64(38.0)

In [33]:
std = np.std(l, ddof=1)   # sample standard deviation, ddof=1 for unbiased estimate
se = std / np.sqrt(n)     # standard error of the mean
se

np.float64(3.998105612017832)

In [34]:
t_stat = (x - mu) / se  # one-sample t-statistic
t_stat

np.float64(0.7503553660469486)

#### Step 4 : Critical Value Method

In [35]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160082949)

**|t| = 0.750 < t_crit = 2.201 → Fail to reject H₀.**

#### Step 5 : p-value Method

In [36]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.4687872850754634)

**p-value = 0.4688 > 0.05 → Fail to reject H₀.**

#### Step 6 : Confidence Interval

In [37]:
moe = t_crit * se      # margin of error
lower = x - moe        # lower bound of 95% CI
upper = x + moe        # upper bound of 95% CI
print(f"95% CI: ({lower:.4f}, {upper:.4f})")

95% CI: (29.2002, 46.7998)


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Sample mean | 38.0 |
| Null mean (μ₀) | 35 |
| t-statistic | 0.7504 |
| df | 11 |
| Critical value (±) | 2.2010 |
| p-value | 0.4688 |
| 95% CI | (29.2002, 46.7998) |
| Decision | Fail to reject H₀ |

The sample mean of 38 is not significantly different from the claimed average of 35 at α = 0.05. The 95% CI spans from 29.2 to 46.8 and comfortably contains 35.

#### Part (b) : Bootstrap Hypothesis Test

#### Why bootstrap

With only 12 observations and that one high value of 67, I'm not super confident about normality. The bootstrap lets me test the same hypothesis without any distributional assumptions. The idea: shift the data so H₀ is exactly true, resample thousands of times, and see how often the bootstrap means wander as far from 35 as my observed mean of 38 did.

In [38]:
np.random.seed(0)              # fix seed for reproducibility
shifted = l - x + mu          # shift data so mean equals mu, preserving spread
print(np.mean(shifted))        # verify shifted mean equals 35

35.0


In [39]:
n_boot = 10000                                              # number of bootstrap resamples
boot_samples = np.random.choice(shifted, size=(n_boot, n), replace=True)  # resample with replacement
boot_means = boot_samples.mean(axis=1)                      # mean of each resample
boot_means.shape

(10000,)

In [40]:
obs_diff = abs(x - mu)                              # observed deviation from null mean
extreme = np.abs(boot_means - mu) >= obs_diff       # bootstrap means at least as extreme
p_val_boot = np.mean(extreme)                        # proportion of extreme samples
p_val_boot

np.float64(0.44)

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Observed mean | 38.0 |
| Null mean (μ₀) | 35 |
| Observed deviation | 3.0 |
| Bootstrap resamples | 10,000 |
| Bootstrap p-value | 0.4400 |
| Decision | Fail to reject H₀ |

The bootstrap test reaches the same conclusion as the t-test. Out of 10,000 resamples from a world where μ = 35, about 44% produced a deviation of 3 or more, confirming the observed mean of 38 is not unusual under H₀.

#### Part (c) : Comparison

In [41]:
print(f"t-test p-value    = {p_val:.4f}")      # parametric result
print(f"bootstrap p-value = {p_val_boot:.4f}")  # non-parametric result

t-test p-value    = 0.4688
bootstrap p-value = 0.4400


### ✅ Part (c) Conclusion

| Method | p-value | Decision |
|---|---|---|
| One-sample t-test | 0.4688 | Fail to reject H₀ |
| Bootstrap test | 0.4400 | Fail to reject H₀ |

The two p-values are close but not identical. They diverge most when the normality assumption of the t-test breaks down, such as with heavy-tailed distributions, strong skew, or small samples with outliers. In this case, n = 12 with one potentially influential observation (67), yet both methods agree. When p-values from the two approaches disagree substantially, the bootstrap result is more trustworthy because it makes no distributional assumption about the population.

#### Part (d) : Outlier Detection and Influence

In [42]:
q1 = np.percentile(l, 25)           # first quartile
q3 = np.percentile(l, 75)           # third quartile
iqr = q3 - q1                       # interquartile range
lower_bound = q1 - 1.5 * iqr        # lower IQR fence
upper_bound = q3 + 1.5 * iqr        # upper IQR fence
print(f"Q1 = {q1}, Q3 = {q3}, IQR = {iqr}")
print(f"Bounds: ({lower_bound:.4f}, {upper_bound:.4f})")

Q1 = 28.75, Q3 = 45.5, IQR = 16.75
Bounds: (3.6250, 70.6250)


In [43]:
outliers = l[(l < lower_bound) | (l > upper_bound)]  # values outside IQR fences
print(f"IQR outliers: {outliers}")
print(f"Most extreme value: {l.max()}")
print(f"Mean without 67: {np.mean(l[l != 67]):.4f}")

IQR outliers: []
Most extreme value: 67
Mean without 67: 35.3636


### ✅ Part (d) Conclusion

| Metric | Value |
|---|---|
| Q1 | 28.75 |
| Q3 | 45.50 |
| IQR | 16.75 |
| Lower fence | 3.6250 |
| Upper fence | 70.6250 |
| Formal IQR outliers | None |
| Most extreme value | 67 |
| Mean without 67 | 35.3636 |

No values fall outside the IQR fences, so 67 is not a formal outlier by this rule. However, it is the most extreme observation and pulls the sample mean from 35.4 (without it) to 38.0 (with it), a shift of 2.6 minutes. That shift is responsible for the apparent deviation from the null mean. In a small sample of 12, a single high observation carries meaningful leverage. Removing 67 would bring the sample mean nearly exactly in line with the claimed value of 35, making H₀ even less rejectable.

---

## 📌 Question 6 : Effect Size vs Statistical Significance

Effect size is everything: two t-tests both yield **p = 0.03**.

| | Test 1 | Test 2 |
|---|---|---|
| n per group | 8 | 800 |
| Mean difference | 5 | 5 |
| Pooled SD | 20 | 20 |

**(a)** Compute Cohen's d for each. Are they the same?  
**(b)** Compute statistical power for each test at α = 0.05.  
**(c)** Articulate precisely what the p-value conflates that Cohen's d separates.

In [44]:
from scipy.stats import nct

mean_diff = 5   # observed mean difference (same for both tests)
sp        = 20  # pooled standard deviation (same for both tests)
alpha     = 0.05

#### Part (a) : Cohen's d

#### Why this measure

A raw mean difference of 5 means nothing without context. Is 5 a big deal or a rounding error? Cohen's d answers that by expressing the difference in units of the pooled SD, giving a standardized measure of effect size. Crucially, it doesn't depend on sample size at all, unlike the p-value.

#### Formula

$$d = \frac{\bar{x}_1 - \bar{x}_2}{s_p}$$

In [45]:
d = mean_diff / sp  # Cohen's d — same for both tests because mean_diff and sp are identical
d

0.25

### ✅ Part (a) Conclusion

| Metric | Test 1 (n = 8) | Test 2 (n = 800) |
|---|---|---|
| Mean difference | 5 | 5 |
| Pooled SD | 20 | 20 |
| **Cohen's d** | **0.25** | **0.25** |

Yes: Cohen's d is identical for both tests (d = 0.25, a small-to-medium effect by Cohen's benchmarks of 0.2 / 0.5 / 0.8). The effect size is a property of the population means and variability, not of the sample size. Sample size affects how *detectable* the effect is, not how *large* it is.

#### Part (b) : Statistical Power

#### Why this matters

Power = P(reject H₀ | H₁ is true). Both tests have the same Cohen's d, so the effect is objectively the same size in both cases. But power also depends on sample size through the SE. As n grows, SE shrinks, the t-statistic grows, and it becomes easier to detect the effect. The NCP (non-centrality parameter) captures both:

$$\text{NCP} = \frac{\bar{x}_1 - \bar{x}_2}{SE} = \frac{\bar{x}_1 - \bar{x}_2}{s_p\sqrt{2/n}}$$

Power is computed as the area of the non-central t-distribution beyond the critical value.

In [46]:
# --- Test 1: n = 8 per group ---
n1   = 8
df1  = 2 * n1 - 2                     # degrees of freedom
se1  = sp * np.sqrt(2 / n1)           # standard error of the mean difference
t_stat1 = mean_diff / se1             # t-statistic under H₁
t_crit1 = t.ppf(1 - alpha / 2, df1)  # two-tailed critical value
ncp1    = t_stat1                     # non-centrality parameter equals t-stat under H₁
power1  = nct.sf(t_crit1, df1, ncp1) + nct.cdf(-t_crit1, df1, ncp1)  # two-tailed power
print(f"Test 1 | n={n1}/group | SE={se1:.4f} | t={t_stat1:.4f} | NCP={ncp1:.4f} | Power={power1:.4f}")

Test 1 | n=8/group | SE=10.0000 | t=0.5000 | NCP=0.5000 | Power=0.0753


In [47]:
# --- Test 2: n = 800 per group ---
n2   = 800
df2  = 2 * n2 - 2
se2  = sp * np.sqrt(2 / n2)
t_stat2 = mean_diff / se2
t_crit2 = t.ppf(1 - alpha / 2, df2)
ncp2    = t_stat2
power2  = nct.sf(t_crit2, df2, ncp2) + nct.cdf(-t_crit2, df2, ncp2)
print(f"Test 2 | n={n2}/group | SE={se2:.4f} | t={t_stat2:.4f} | NCP={ncp2:.4f} | Power={power2:.4f}")

Test 2 | n=800/group | SE=1.0000 | t=5.0000 | NCP=5.0000 | Power=0.9988


### ✅ Part (b) Conclusion

| Metric | Test 1 (n = 8/group) | Test 2 (n = 800/group) |
|---|---|---|
| SE | 10.0000 | 1.0000 |
| t-statistic | 0.5000 | 5.0000 |
| df | 14 | 1598 |
| Critical value (±) | 2.1448 | 1.9614 |
| NCP | 0.5000 | 5.0000 |
| **Power** | **0.0753** | **0.9988** |

Same effect size (d = 0.25), same mean difference, same pooled SD: yet power jumps from 7.5% to 99.9%. Test 1 is almost completely blind to the effect; Test 2 will detect it virtually every time. The difference is entirely driven by sample size through the SE: n = 800 shrinks SE by a factor of 10, which multiplies the t-statistic by 10.

#### Part (c) : What the p-value Conflates That Cohen's d Separates

#### The key insight

The t-statistic is literally a product of two things:

$$t = \underbrace{\frac{\bar{x}_1 - \bar{x}_2}{s_p}}_{\text{effect size (Cohen's d)}} \times \underbrace{\sqrt{\frac{n}{2}}}_{\text{sample size term}}$$

The p-value just converts this single number to a probability. It can't separate which factor pushed t up. Cohen's d keeps only the first factor, so it purely measures the signal-to-noise ratio in the population, with no n involved.

**The p-value conflates both. Cohen's d separates them.**

| Scenario | p-value says | Cohen's d says | Reality |
|---|---|---|---|
| Large n, tiny effect | Significant (p < 0.05) | Small (d ≈ 0.01) | Statistically significant, practically irrelevant |
| Small n, large effect | Not significant (p > 0.05) | Large (d ≈ 0.8) | Practically important, underpowered study |
| Large n, large effect | Highly significant | Large | Both agree |

This is why modern statistics requires reporting effect sizes alongside p-values: they answer different questions.

### ✅ Part (c) Conclusion

| Quantity | What it measures | Depends on n? | Answers |
|---|---|---|---|
| p-value | Rarity of data under H₀ | Yes | Is the result surprising? |
| Cohen's d | Standardised effect magnitude | No | Is the effect large? |
| Power | Probability of detection | Yes | Would we find it again? |

The p-value conflates **effect magnitude** and **sample size** into one number. A large n can produce p < 0.05 even for a trivially small d = 0.01. Cohen's d strips out sample size entirely and reports only the signal-to-noise ratio in the population. In this question both tests share d = 0.25, but the p-values and power differ dramatically: a direct demonstration that p-values should never be read as effect size measures.

---

## 📌 Question 7 : Permutation Test

Salaries (in $k): Engineering [95, 102, 88, 110, 98, 105, 92] vs Marketing [78, 82, 75, 91, 85, 79].

**(a)** Run a standard two-sample t-test (Welch's).
**(b)** Run a permutation test with 5,000 permutations.
**(c)** With n = 7 and n = 6, which test is more valid and why?
**(d)** What does "the permutation test is assumption-free" actually mean?

#### Step 1 : Hypotheses

- **H₀ :** μ_eng = μ_mkt (Engineering and Marketing have the same mean salary)
- **H₁ :** μ_eng ≠ μ_mkt (the two group means are different)

Two-tailed test at α = 0.05.

#### Step 2 : Why both tests

With only n = 7 and n = 6, I'm dealing with very small groups. The t-test technically works here but relies on an assumption that the underlying data is roughly normal. With so few observations I can't really verify that. The permutation test sidesteps this entirely: it only assumes that under H₀, group labels are exchangeable. That's a much weaker assumption, and one that's reasonable for salary data from the same company.

#### Part (a) : Two-Sample Welch's t-Test

#### Step 3 : Test Statistic

$$t = \frac{\bar{x}_1 - \bar{x}_2}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}}$$

In [48]:
eng = np.array([95, 102, 88, 110, 98, 105, 92])  # Engineering salaries
mkt = np.array([78, 82, 75, 91, 85, 79])            # Marketing salaries

n_eng = len(eng)   # Engineering group size
n_mkt = len(mkt)   # Marketing group size

x_eng = np.mean(eng)  # Engineering sample mean
x_mkt = np.mean(mkt)  # Marketing sample mean

x_eng, x_mkt

(np.float64(98.57142857142857), np.float64(81.66666666666667))

In [49]:
obs_diff = x_eng - x_mkt  # observed difference in group means
obs_diff

np.float64(16.904761904761898)

In [50]:
var_eng = np.var(eng, ddof=1)  # Engineering sample variance
var_mkt = np.var(mkt, ddof=1)  # Marketing sample variance

se = np.sqrt(var_eng / n_eng + var_mkt / n_mkt)  # Welch standard error
se

np.float64(3.7173369640521416)

In [51]:
t_stat = obs_diff / se  # Welch t-statistic
t_stat

np.float64(4.547546286020462)

#### Step 4 : Critical Value Method

In [52]:
alpha = 0.05

# Welch-Satterthwaite degrees of freedom
num = (var_eng / n_eng + var_mkt / n_mkt) ** 2
den = (var_eng / n_eng) ** 2 / (n_eng - 1) + (var_mkt / n_mkt) ** 2 / (n_mkt - 1)
df_welch = num / den

t_crit = t.ppf(1 - alpha / 2, df_welch)  # two-tailed critical value
t_crit

np.float64(2.204962782820248)

**|t| > t_crit → Reject H₀.** 

#### Step 5 : p-value Method

In [53]:
p_val = 2 * t.sf(abs(t_stat), df_welch)  # two-tailed Welch p-value
p_val

np.float64(0.0008651210641434828)

#### Step 6 : Confidence Interval

In [54]:
margin = t_crit * se              # margin of error
lo = obs_diff - margin             # lower bound of 95% CI
hi = obs_diff + margin             # upper bound of 95% CI
print(f"95% CI for (μ_eng - μ_mkt): ({lo:.4f}, {hi:.4f})")

95% CI for (μ_eng - μ_mkt): (8.7082, 25.1014)


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Engineering mean | 98.57 |
| Marketing mean | 81.67 |
| Observed difference | 16.90 |
| t-statistic | ~4.5475 |
| df (Welch) | ~10 |
| p-value | < 0.05 |
| Decision | Reject H₀ |

The t-test finds a statistically significant difference in mean salaries. Engineering earns more on average, and the result is significant at α = 0.05.

#### Part (b) : Permutation Test

The permutation test asks: if salary has nothing to do with department (H₀ is true), then the group labels (Eng vs Mkt) are arbitrary. I can randomly shuffle them and recompute the mean difference. If I do this 5,000 times, the distribution of shuffled differences tells me how extreme the observed difference of 16.9 actually is.

In [55]:
np.random.seed(42)  # fix seed for reproducibility

combined = np.concatenate([eng, mkt])  # pool all salaries together
n_total  = len(combined)               # total observations
n_perm   = 5000                        # number of permutations

perm_diffs = np.zeros(n_perm)  # store permuted mean differences

for i in range(n_perm):
    shuffled  = np.random.permutation(combined)    # randomly reassign all labels
    perm_eng  = shuffled[:n_eng]                   # first n_eng go to "Engineering"
    perm_mkt  = shuffled[n_eng:]                   # rest go to "Marketing"
    perm_diffs[i] = np.mean(perm_eng) - np.mean(perm_mkt)  # permuted difference

perm_diffs.shape

(5000,)

In [56]:
# two-tailed: count how many permuted diffs are at least as extreme as observed
p_val_perm = np.mean(np.abs(perm_diffs) >= np.abs(obs_diff))
p_val_perm

np.float64(0.0014)

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| Observed mean difference | 16.90 |
| Permutations | 5,000 |
| Permutation p-value | < 0.05 |
| Decision | Reject H₀ |

Out of 5,000 random label shuffles, very few produce a mean difference as large as 16.9. The permutation p-value confirms that the observed separation between groups is unlikely to be just random chance.

#### Part (c) : Comparison

#### Step 7 : Side-by-side

In [57]:
print(f"Welch t-test p-value   = {p_val:.4f}")
print(f"Permutation p-value    = {p_val_perm:.4f}")
print(f"Observed difference    = {obs_diff:.4f}")
print(f"Welch df               = {df_welch:.2f}")

Welch t-test p-value   = 0.0009
Permutation p-value    = 0.0014
Observed difference    = 16.9048
Welch df               = 10.84


### ✅ Part (c) Conclusion

| Method | p-value | Decision |
|---|---|---|
| Welch t-test | < 0.05 | Reject H₀ |
| Permutation test | < 0.05 | Reject H₀ |

Both tests agree. With n = 7 and n = 6, the permutation test is more trustworthy. The Welch t-test assumes the salary data within each group is approximately normal. With only 6-7 observations per group, I simply can't verify that. The permutation test makes no such assumption: it only requires that labels be exchangeable under H₀, which is a reasonable baseline assumption when testing for a group difference.

#### Part (d) : What "Assumption-Free" Actually Means

Calling the permutation test "assumption-free" is a slight exaggeration. What it really means:

**What the permutation test does NOT assume:**
- No normality required within groups
- No specific distributional family (t, normal, exponential, etc.)
- No assumption about equal or unequal variances

**What the permutation test DOES assume:**
- **Exchangeability under H₀:** If salary is truly unrelated to department, then any permutation of labels is just as likely as any other. This is the core assumption.
- **Independence of observations:** Salaries are from different people and assumed independent.
- **Same distribution under H₀:** Both groups are drawn from the same distribution when H₀ is true (otherwise the labels wouldn't be truly exchangeable).

| Assumption | t-test | Permutation test |
|---|---|---|
| Normality within groups | Yes | No |
| Specific distribution family | Yes (t) | No |
| Equal/unequal variances (choose one) | Yes | No |
| Independence of observations | Yes | Yes |
| Exchangeability of labels under H₀ | Implicit | Yes (explicit) |

So the permutation test trades strong parametric assumptions (normality, t-distribution) for one weaker assumption (exchangeability). That's a good trade, especially at small sample sizes. The tradeoff is computational: 5,000 permutations instead of a formula. With modern hardware that's not a real concern.

### ✅ Part (d) Conclusion

"Assumption-free" really means: free from distributional assumptions, not free from all assumptions. The permutation test's core assumption (exchangeability under H₀) is almost always reasonable when testing whether group membership matters. For small samples like n = 7 and n = 6, it's the more honest and robust choice.

## 📌 Question 8 : Bayesian t-Test

Bayesian t-test: n = 12 observations, x̄ = 8.5, s = 2.1. H₀: μ = 7.

**(a)** Run the frequentist one-sample t-test.
**(b)** Using the JZS Cauchy prior (scale = 0.707), compute the Bayes Factor BF₁₀.
**(c)** Interpret: what do BF = 5, BF = 0.2, and BF = 0.03 each mean?
**(d)** Name one thing a Bayes Factor can do that a p-value fundamentally cannot.

In [58]:
n  = 12    # sample size
x  = 8.5   # sample mean
s  = 2.1   # sample standard deviation
mu = 7     # null hypothesis mean
alpha = 0.05

#### Part (a) : Frequentist One-Sample t-Test

#### Step 1 : Hypotheses

- **H₀ :** μ = 7 (true mean equals 7)
- **H₁ :** μ ≠ 7 (true mean differs from 7)

Two-tailed test at α = 0.05.

#### Step 2 : Why this test

One sample, unknown σ, small n = 12. The population SD is unknown so a Z-test is off the table. With only 12 observations I use the t-distribution, which has heavier tails to account for the extra uncertainty in estimating σ from a small sample.

#### Step 3 : Test Statistic

$$t = \frac{\bar{x} - \mu_0}{s / \sqrt{n}}$$

Degrees of freedom: df = n − 1 = 11

In [59]:
df = n - 1                  # degrees of freedom
se = s / n**0.5             # standard error of the mean
t_stat = (x - mu) / se     # one-sample t-statistic
t_stat

2.4743582965269675

#### Step 4 : Critical Value Method

In [60]:
t_crit = t.ppf(1 - alpha / 2, df)  # two-tailed critical value at alpha=0.05, df=11
t_crit

np.float64(2.200985160082949)

**|t| = 2.481 > t_crit = 2.201 → Reject H₀.**

#### Step 5 : p-value Method

In [61]:
p_val = 2 * t.sf(abs(t_stat), df)  # two-tailed p-value
p_val

np.float64(0.030881561145156815)

**p-value = 0.0311 < 0.05 → Reject H₀.**

#### Step 6 : Confidence Interval

In [62]:
se_val = s / np.sqrt(n)
lo = x - t_crit * se_val     # lower bound of 95% CI
hi = x + t_crit * se_val     # upper bound of 95% CI
print(f"95% CI: ({lo:.4f}, {hi:.4f})")

95% CI: (7.1657, 9.8343)


### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| Sample mean | 8.5 |
| Null mean (μ₀) | 7 |
| t-statistic | 2.4807 |
| df | 11 |
| Critical value (±) | 2.2010 |
| p-value | 0.0311 |
| 95% CI | (7.1640, 9.8360) |
| Decision | Reject H₀ |

The sample mean of 8.5 is significantly different from the null value of 7 at α = 0.05. The 95% confidence interval lies entirely above 7, and the p-value of 0.031 falls below the significance threshold.

#### Part (b) : Bayes Factor BF₁₀ via JZS Cauchy Prior

#### Why the JZS prior

The JZS (Jeffreys-Zellner-Siow) prior places a Cauchy distribution on the standardised effect size δ = (μ − μ₀)/σ, with scale r = 0.707 (= 1/√2). This is a default prior: it's symmetric around zero (no a priori direction assumed), has heavy tails (allows large effects), and is invariant under rescaling. The Bayes Factor BF₁₀ is the ratio of marginal likelihoods P(data | H₁) / P(data | H₀). Computing it requires integrating the Cauchy prior over all possible effect sizes.

#### Formula

$$BF_{10} = \frac{\int_0^\infty \left(1 + \frac{n\delta^2}{t^2 + n - 1}\right)^{-n/2} \cdot p(\delta \mid r) \, d\delta}{(1 + t^2/(n-1))^{-n/2}}$$

In practice this integral is computed numerically. The `pingouin` library provides a direct implementation.

In [63]:
from scipy.integrate import quad

r = 1 / np.sqrt(2)

BF10, _ = quad(
    lambda g: (
        g**(-3/2) * np.exp(-1/(2*g)) / np.sqrt(2*np.pi) *
        (1 + n*g*r**2)**(-1/2) *
        (1 + t_stat**2 / (df*(1 + n*g*r**2)))**(-(df+1)/2) /
        (1 + t_stat**2 / df)**(-(df+1)/2)
    ),
    0, np.inf
)

print(f"BF₁₀ = {BF10:.4f}")

BF₁₀ = 2.4069


### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| t-statistic | 2.4744 |
| df | 11 |
| n | 12 |
| JZS prior scale (r) | 0.707 |
| **BF₁₀** | **≈ 2.41** |

BF₁₀ ≈ 2.41 means the data are about 2.4 times more likely under H₁ than under H₀. By Jeffreys' scale this constitutes **anecdotal evidence** in favour of H₁ (BF between 1 and 3). The frequentist test rejected H₀ at α = 0.05, but the Bayes Factor tells a more cautious story: there is only weak-to-moderate support for the alternative.

#### Part (c) : Interpreting Specific Bayes Factors

#### Jeffreys' Interpretation Scale

| BF₁₀ | Interpretation |
|---|---|
| > 100 | Extreme evidence for H₁ |
| 30–100 | Very strong evidence for H₁ |
| 10–30 | Strong evidence for H₁ |
| 3–10 | Moderate evidence for H₁ |
| 1–3 | Anecdotal evidence for H₁ |
| 1 | No evidence either way |
| 1/3–1 | Anecdotal evidence for H₀ |
| 1/10–1/3 | Moderate evidence for H₀ |
| < 1/10 | Strong evidence for H₀ |

**BF₁₀ = 5** → The data are 5 times more likely under H₁ than H₀. This is moderate evidence in favour of H₁. Worth noting, but a replication would strengthen the conclusion.

**BF₁₀ = 0.2 = 1/5** → The data are 5 times more likely under H₀ than H₁. This is moderate evidence in favour of H₀ — the data are actively supporting the null, not just failing to reject it.

**BF₁₀ = 0.03 ≈ 1/33** → The data are approximately 33 times more likely under H₀ than H₁. This is strong-to-very-strong evidence for H₀. The null hypothesis is strongly supported.

### ✅ Part (c) Conclusion

| BF₁₀ | Numeric meaning | Verbal interpretation |
|---|---|---|
| 5 | H₁ is 5× more likely | Moderate evidence for H₁ |
| 0.2 | H₀ is 5× more likely | Moderate evidence for H₀ |
| 0.03 | H₀ is ~33× more likely | Strong evidence for H₀ |

Unlike the p-value, Bayes Factors are symmetric: they can support either hypothesis, and their magnitude quantifies the strength of that support on a continuous scale.

#### Part (d) : What a Bayes Factor Can Do That a p-value Cannot

**A Bayes Factor can quantify evidence in favour of the null hypothesis. A p-value fundamentally cannot.**

A p-value answers: *"How surprising is this data if H₀ is true?"* A large p-value (e.g. p = 0.8) does not mean H₀ is supported — it just means the data are not inconsistent with H₀. Failing to reject is not evidence for the null; it could simply reflect low power, a small sample, or high variance.

A Bayes Factor answers: *"How much more likely is the data under H₁ versus H₀?"* When BF₁₀ = 0.03, the data are actively supporting H₀ — 33 times more likely under H₀ than H₁. That's a positive, quantifiable statement about the null, not a silence.

### ✅ Part (d) Conclusion

| | p-value | Bayes Factor |
|---|---|---|
| Can reject H₁ | Yes (implicitly, via rejection) | Yes |
| Can support H₀ | **No** | **Yes** |
| Quantifies evidence strength | No | Yes |
| Depends on stopping rule | Yes | No |
| Requires prior | No | Yes |

This is the single most important practical difference. In fields where demonstrating that an effect is absent matters (e.g. safety testing, bioequivalence, cognitive neuroscience null results), the Bayes Factor is the appropriate tool. The p-value cannot distinguish "no evidence for an effect" from "evidence for no effect" — the Bayes Factor can.

## 📌 Question 9 : Hotelling's T² (Multivariate Extension)

20 patients were measured on **both** blood pressure and cholesterol before/after treatment.

Mean differences: d̄_BP = 8.2, d̄_Chol = 12.1  
Covariance matrix of differences: **S = [[36, 18], [18, 64]]**

**(a)** Compute T² = n · d̄ᵀ · S⁻¹ · d̄  
**(b)** Convert to F: F = T²·(n − p) / [p·(n − 1)]  
**(c)** Why is this better than running two separate paired t-tests?

#### Step 1 : Hypotheses

- **H₀ :** μ_BP = 0 and μ_Chol = 0 simultaneously (treatment has no effect on either outcome)
- **H₁ :** At least one mean difference is non-zero (treatment affects at least one outcome)

Two-tailed (omnibus) test at α = 0.05.

#### Step 2 : Why this test

Blood pressure and cholesterol are correlated outcomes measured on the same patients. Running two separate paired t-tests ignores that correlation and inflates the Type I error rate. Hotelling's T² is the multivariate generalisation of the paired t-test: it tests both outcomes jointly in a single test, respects the covariance structure between them, and controls the overall α at 0.05.

#### Part (a) : Computing T²

#### Why this formula

In the univariate paired t-test, t² = n · (d̄ − 0)² / s² = n · d̄ · (1/s²) · d̄. Hotelling's T² is exactly that idea extended to vectors: instead of dividing by the scalar variance, we multiply by the inverse of the covariance matrix S. S⁻¹ plays the role of 1/s² and simultaneously accounts for the variance of each variable and the covariance between them.

$$T^2 = n \cdot \bar{\mathbf{d}}^\top \cdot S^{-1} \cdot \bar{\mathbf{d}}$$

In [ ]:
from scipy.stats import f
import numpy as np

n     = 20                                    # number of patients
p     = 2                                     # number of outcome variables
d_bar = np.array([8.2, 12.1])                # vector of mean differences [BP, Chol]
S     = np.array([[36, 18], [18, 64]])        # covariance matrix of differences
S_inv = np.linalg.inv(S)                     # inverse of covariance matrix
alpha = 0.05                                  # significance level

T2 = n * d_bar @ S_inv @ d_bar               # Hotelling's T² statistic
T2

### ✅ Part (a) Conclusion

| Metric | Value |
|---|---|
| n | 20 |
| d̄ | [8.2, 12.1] |
| S | [[36, 18], [18, 64]] |
| S⁻¹ (approx) | [[0.0308, −0.0087], [−0.0087, 0.0173]] |
| **T²** | **60.628** |

The T² statistic summarises how far the observed mean difference vector is from the origin (zero vector), weighted by the inverse covariance matrix. On its own it is not directly interpretable without converting to F.

#### Part (b) : Converting T² to an F-Statistic

#### Why convert to F

T² follows a scaled F-distribution under H₀, not an F-distribution directly. The exact conversion is:

$$F = \frac{T^2 \cdot (n - p)}{p \cdot (n - 1)}$$

This F-statistic has degrees of freedom df₁ = p and df₂ = n − p, and we can look up its p-value from the standard F-distribution. This conversion is exact, not an approximation.

In [ ]:
f_stat = T2 * (n - p) / (p * (n - 1))     # convert T² to F-statistic
f_stat

#### Step 4 : Critical Value Method

In [ ]:
df1    = p                                  # numerator df
df2    = n - p                              # denominator df
f_crit = f.ppf(1 - alpha, df1, df2)        # one-tailed upper critical value (F is always right-tailed)
f_crit

**F = 28.719 > F_crit = 3.555 → Reject H₀.**

#### Step 5 : p-value Method

In [ ]:
p_val = f.sf(f_stat, df1, df2)             # p-value from F-distribution
print(f"F-statistic : {f_stat:.4f}")
print(f"F_crit      : {f_crit:.4f}")
print(f"p-value     : {p_val:.8f}")

**p-value = 0.00000251 < 0.05 → Reject H₀.**

### ✅ Part (b) Conclusion

| Metric | Value |
|---|---|
| T² | 60.628 |
| F-statistic | 28.719 |
| df₁ | 2 |
| df₂ | 18 |
| F_crit (α = 0.05) | 3.555 |
| p-value | 0.00000251 |
| Decision | Reject H₀ |

The F-statistic of 28.72 far exceeds the critical value of 3.56, and the p-value is vanishingly small. There is very strong evidence that treatment produced a significant simultaneous change in at least one of the two outcomes.

#### Part (c) : Why Hotelling's T² Is Better Than Two Separate t-Tests

In [ ]:
from scipy.stats import t as t_dist

# Individual paired t-tests using diagonal entries of S as variances
se_BP   = np.sqrt(36) / np.sqrt(n)                         # SE for blood pressure
se_Chol = np.sqrt(64) / np.sqrt(n)                         # SE for cholesterol

t_BP    = 8.2  / se_BP                                     # t-statistic for BP
t_Chol  = 12.1 / se_Chol                                   # t-statistic for cholesterol

p_BP    = 2 * t_dist.sf(abs(t_BP),   n - 1)               # two-tailed p-value for BP
p_Chol  = 2 * t_dist.sf(abs(t_Chol), n - 1)               # two-tailed p-value for cholesterol

corr_BP_Chol = 18 / (np.sqrt(36) * np.sqrt(64))            # correlation between the two outcomes

print(f"Individual t_BP   = {t_BP:.4f},  p = {p_BP:.6f}")
print(f"Individual t_Chol = {t_Chol:.4f},  p = {p_Chol:.6f}")
print(f"Bonferroni-adjusted alpha = {alpha / p:.3f}")
print(f"Correlation between outcomes: {corr_BP_Chol:.4f}")

### ✅ Part (c) Conclusion

| | Two Separate t-Tests | Hotelling's T² |
|---|---|---|
| Number of tests | 2 | 1 |
| Nominal α | 0.05 each | 0.05 overall |
| Actual Type I error (two tests, r = 0.375) | > 0.05 | Exactly 0.05 |
| Uses correlation between outcomes | No | Yes |
| Single omnibus decision | No | Yes |
| p (BP) | 0.000007 | |
| p (Chol) | 0.000002 | |
| Combined p-value | N/A | 0.00000251 |

Running two separate paired t-tests has three problems. First, it inflates the Type I error rate: with two tests at α = 0.05 each, the probability of at least one false positive is no longer 5%. Bonferroni correction adjusts to α = 0.025 per test, but that is conservative and loses power. Second, the tests treat blood pressure and cholesterol as if they are independent, ignoring the covariance of 18 (correlation ≈ 0.375). Hotelling's T² accounts for that covariance through S⁻¹. Third, two separate tests give two separate decisions with no coherent overall conclusion. T² provides a single multivariate decision that correctly controls the family-wise error rate at α = 0.05 and uses all the information in the covariance structure.

In this particular dataset the individual p-values are both tiny, so the inflation issue does not change the conclusion. But in borderline cases, ignoring the covariance can lead to contradictory results across the two tests, exactly the situation Hotelling's T² is designed to prevent.